protocols/kad/src/query/peers/closest/disjoint.rs

```diff
@@ -0,0 +1,965 @@
+    where
+        I: IntoIterator<Item = PeerId>
+    {
+        for peer in peers {
+            if let Some(PeerState{ initiated_by, .. }) = self.contacted_peers.get_mut(&peer) {
+                self.iters[*initiated_by].finish();
+            }
+        }
+    }
+
+    /// Immediately transitions the iterator to [`PeersIterState::Finished`].
+    pub fn finish(&mut self) {
+        for iter in &mut self.iters {
+            iter.finish();
+        }
+    }
+
+    /// Checks whether the iterator has finished.
+    pub fn is_finished(&self) -> bool {
+        for iter in &self.iters {
+            if iter.is_finished() {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Note: In the case of no adversarial peers or connectivity issues along
+    ///       any path, all paths return the same result, deduplicated through
+    ///       the `ResultIter`, thus overall `into_result` returns
+    ///       `num_results`. In the case of adversarial peers or connectivity
+    ///       issues `ClosestDisjointPeersIter` tries to return the
+    ///       `num_results` closest benign peers, but as it can not
+    ///       differentiate benign from faulty paths it as well returns faulty
+    ///       peers and thus overall returns more than `num_results` peers.
+    pub fn into_result(self) -> impl Iterator<Item = PeerId> {
+        let result_per_path= self.iters.into_iter()
+            .map(|iter| iter.into_result().map(Key::from));
+
+        ResultIter::new(self.target, result_per_path).map(Key::into_preimage)
+    }
+}
+
+/// Index into the [`ClosestDisjointPeersIter`] `iters` vector.
+#[derive(Copy, Clone, Debug)]
+struct IteratorIndex(usize);
+
+impl From<usize> for IteratorIndex {
+    fn from(i: usize) -> Self {
+        IteratorIndex(i)
+    }
+}
+
+impl From<IteratorIndex> for usize {
+    fn from(i: IteratorIndex) -> Self {
+        i.0
+    }
+}
+
+impl Add<usize> for IteratorIndex {
+    type Output = Self;
+
+    fn add(self, rhs: usize) -> Self::Output {
+        (self.0 + rhs).into()
+    }
+}
+
+impl Index<IteratorIndex> for Vec<ClosestPeersIter> {
+    type Output = ClosestPeersIter;
+
+    fn index(&self, index: IteratorIndex) -> &Self::Output {
+        &self[index.0]
+    }
+}
+
+impl IndexMut<IteratorIndex> for Vec<ClosestPeersIter> {
+    fn index_mut(&mut self, index: IteratorIndex) -> &mut Self::Output {
+        &mut self[index.0]
+    }
+}
+
+/// State tracking the iterator that yielded (i.e. tried to contact) a peer. See
+/// [`ClosestDisjointPeersIter::on_success`] for details.
+#[derive(Debug)]
+struct PeerState {
+    /// First iterator to yield the peer. Will be notified both of the outcome
+    /// (success/failure) as well as the closer peers.
+    initiated_by: IteratorIndex,
+    /// Keeping track of the response state. In case other iterators later on
+    /// yield the same peer, they can be notified of the response outcome.
+    response: ResponseState,
```

Comment:
As mentioned in the other comments, it would be great if we could remove this field, though it may be hard to do given the implementation requirements on `next`, but at least we really need to make sure there is never a state mismatch between `response` and the state according to the iterator identified by `initiated_by`, which can currently happen.